lib/rules/logical-assignment-operators.js

```diff
@@ -0,0 +1,457 @@
+
+                                // -> foo = foo || (bar)
+                                yield ruleFixer.insertTextBefore(assignment.right, "(");
+                                yield ruleFixer.insertTextAfter(assignment.right, ")");
+                            }
+                        }
+                    };
+
+                    context.report(createConditionalFixer(descriptor, suggestion, canBeFixed(assignment.left)));
+                }
+            };
+        }
+
+        return {
+
+            // foo = foo || bar
+            "AssignmentExpression[operator='='][right.type='LogicalExpression']"(assignment) {
+                if (!astUtils.isSameReference(assignment.left, assignment.right.left, true)) {
+                    return;
+                }
+
+                const descriptor = {
+                    messageId: "assignment",
+                    node: assignment,
+                    data: { operator: assignment.right.operator }
+                };
+                const suggestion = {
+                    messageId: "useLogicalOperator",
+                    data: { operator: assignment.operator },
+                    *fix(ruleFixer) {
+                        if (sourceCode.getCommentsInside(assignment).length > 0) {
+                            return;
+                        }
+
+                        // No need for parenthesis around the assignment based on precedence as the precedence stays the same even with changed operator
+                        const assignmentOperatorToken = getOperatorToken(assignment);
+
+                        // -> foo ||= foo || bar
+                        yield ruleFixer.insertTextBefore(assignmentOperatorToken, assignment.right.operator);
+
+                        // -> foo ||= bar
+                        yield ruleFixer.removeRange([assignment.right.range[0], assignment.right.right.range[0]]);
+
+                        if (astUtils.isParenthesised(sourceCode, assignment.right.right)) {
+
+                            // Opening parenthesis already removed
+                            const closingParenthesis = sourceCode.getTokenAfter(assignment.right.right);
+
+                            yield ruleFixer.remove(closingParenthesis);
+                        }
+                    }
+                };
+
+                context.report(createConditionalFixer(descriptor, suggestion, canBeFixed(assignment.left)));
+            },
+
+            // foo || (foo = bar)
+            'LogicalExpression[right.type="AssignmentExpression"][right.operator="="]'(logical) {
+                if (astUtils.isSameReference(logical.left, logical.right.left)) {
+                    context.report({
+                        messageId: "logical",
+                        node: logical,
+                        data: { operator: logical.operator },
+                        *fix(ruleFixer) {
```

Comment:
This should be a suggestion if the reference is more complex than a simple member expression (similar to the `if` case).

```js
a.b.c || (a.b.c = foo) // triggers `a.b` twice if `a.b.c` is falsy

a.b.c ||= foo // always triggers `a.b` exactly once
```